llmfoundry/utils/builders.py

```diff
@@ -505,7 +505,11 @@ def _validate_cfg(icl_cfg: DictConfig):
             if dist.get_local_rank() == 0 and os.path.exists(destination_path):
                 os.remove(destination_path)
             dist.barrier()
-
+            early_stopping_criteria = icl_cfg.get('early_stopping_criteria',
+                                                  None)
+            early_stopping_criteria = list(
```

Comment:
why is the list cast necessary?

Reply:
I found a difficult to comprehend error when I didn't include it. I'm not actually sure of the root cause, but this is the stack trace.


```
******************************
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/jeremy.dohmann/LLMWork/llm-foundry/scripts/eval/eval.py:507 in <module>                   │
│                                                                                                  │
│   504 │   cli_cfg = om.from_cli(args_list)                                                       │
│   505 │   cfg = om.merge(yaml_cfg, cli_cfg)                                                      │
│   506 │   assert isinstance(cfg, DictConfig)                                                     │
│ ❱ 507 │   main(cfg)                                                                              │
│   508                                                                                            │
│                                                                                                  │
│ /Users/jeremy.dohmann/LLMWork/llm-foundry/scripts/eval/eval.py:351 in main                       │
│                                                                                                  │
│   348 │   trainers = []                                                                          │
│   349 │   for model_cfg in model_configs:                                                        │
│   350 │   │   (trainer, logger_keys, eval_gauntlet_callback,                                     │
│ ❱ 351 │   │    eval_gauntlet_df) = evaluate_model(                                               │
│   352 │   │   │    model_cfg=model_cfg,                                                          │
│   353 │   │   │    dist_timeout=dist_timeout,                                                    │
│   354 │   │   │    run_name=run_name,                                                            │
│                                                                                                  │
│ /Users/jeremy.dohmann/LLMWork/llm-foundry/scripts/eval/eval.py:230 in evaluate_model             │
│                                                                                                  │
│   227 │   if torch.cuda.is_available():                                                          │
│   228 │   │   torch.cuda.synchronize()                                                           │
│   229 │   a = time.time()                                                                        │
│ ❱ 230 │   trainer.eval(eval_dataloader=evaluators,                                               │
│   231 │   │   │   │    subset_num_batches=eval_subset_num_batches)                               │
│   232 │   if torch.cuda.is_available():                                                          │
│   233 │   │   torch.cuda.synchronize()                                                           │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/composer/trainer/trai │
│ ner.py:2844 in eval                                                                              │
│                                                                                                  │
│   2841 │   │                                                                                     │
│   2842 │   │   for evaluator in evaluators:                                                      │
│   2843 │   │   │   eval_subset_num_batches = evaluator.subset_num_batches if subset_num_batches  │
│ ❱ 2844 │   │   │   self._eval_loop(                                                              │
│   2845 │   │   │   │   evaluator=evaluator,                                                      │
│   2846 │   │   │   │   metrics=self.state.eval_metrics[evaluator.label],                         │
│   2847 │   │   │   │   subset_num_batches=eval_subset_num_batches,                               │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/composer/trainer/trai │
│ ner.py:2921 in _eval_loop                                                                        │
│                                                                                                  │
│   2918 │   │   │   │   │   │   │   │   │     'DistributedSampler to correctly drop duplicate sa  │
│   2919 │   │   │                                                                                 │
│   2920 │   │   │   for self.state.batch in self._iter_dataloader(TrainerMode.EVAL):              │
│ ❱ 2921 │   │   │   │   self.state.batch = self.state.device.batch_to_device(self.state.batch)    │
│   2922 │   │   │   │   self.state.batch = data_spec.device_transforms(self.state.batch)          │
│   2923 │   │   │   │                                                                             │
│   2924 │   │   │   │   # Count the batch size and num tokens before any events run               │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/composer/devices/devi │
│ ce.py:78 in batch_to_device                                                                      │
│                                                                                                  │
│    75 │   │   │   │   return self.tensor_to_device(x)                                            │
│    76 │   │   │   return x                                                                       │
│    77 │   │                                                                                      │
│ ❱  78 │   │   return _map_batch(batch, _to_device)                                               │
│    79 │                                                                                          │
│    80 │   def optimizer_to_device(self, optimizer: Optimizer) -> Optimizer:                      │
│    81 │   │   """Invoked by the :class:`.Trainer` to move the optimizer's state onto the devic   │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/composer/devices/devi │
│ ce.py:108 in _map_batch                                                                          │
│                                                                                                  │
│   105 │   │   The type of ``batch`` is preserved.                                                │
│   106 │   """                                                                                    │
│   107 │   if isinstance(batch, Mapping):                                                         │
│ ❱ 108 │   │   return {k: _map_batch(v, map_fn) for k, v in batch.items()}                        │
│   109 │                                                                                          │
│   110 │   if isinstance(batch, Sequence) and not isinstance(batch, (str, bytes)):                │
│   111 │   │   try:                                                                               │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/composer/devices/devi │
│ ce.py:108 in <dictcomp>                                                                          │
│                                                                                                  │
│   105 │   │   The type of ``batch`` is preserved.                                                │
│   106 │   """                                                                                    │
│   107 │   if isinstance(batch, Mapping):                                                         │
│ ❱ 108 │   │   return {k: _map_batch(v, map_fn) for k, v in batch.items()}                        │
│   109 │                                                                                          │
│   110 │   if isinstance(batch, Sequence) and not isinstance(batch, (str, bytes)):                │
│   111 │   │   try:                                                                               │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/composer/devices/devi │
│ ce.py:112 in _map_batch                                                                          │
│                                                                                                  │
│   109 │                                                                                          │
│   110 │   if isinstance(batch, Sequence) and not isinstance(batch, (str, bytes)):                │
│   111 │   │   try:                                                                               │
│ ❱ 112 │   │   │   return type(batch)(_map_batch(x, map_fn) for x in batch)  # type: ignore       │
│   113 │   │   except TypeError:                                                                  │
│   114 │   │   │   return [_map_batch(x, map_fn) for x in batch]                                  │
│   115 │   return map_fn(batch)                                                                   │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/omegaconf/listconfig. │
│ py:82 in __init__                                                                                │
│                                                                                                  │
│    79 │   │   │   │   self.__dict__["_metadata"] = metadata                                      │
│    80 │   │   │   self._set_value(value=content, flags=flags)                                    │
│    81 │   │   except Exception as ex:                                                            │
│ ❱  82 │   │   │   format_and_raise(node=None, key=key, value=None, cause=ex, msg=str(ex))        │
│    83 │                                                                                          │
│    84 │   def _validate_get(self, key: Any, value: Any = None) -> None:                          │
│    85 │   │   if not isinstance(key, (int, slice)):                                              │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/omegaconf/_utils.py:8 │
│ 99 in format_and_raise                                                                           │
│                                                                                                  │
│    896 │   │   ex.ref_type = ref_type                                                            │
│    897 │   │   ex.ref_type_str = ref_type_str                                                    │
│    898 │                                                                                         │
│ ❱  899 │   _raise(ex, cause)                                                                     │
│    900                                                                                           │
│    901                                                                                           │
│    902 def type_str(t: Any, include_module_name: bool = False) -> str:                           │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/omegaconf/_utils.py:7 │
│ 97 in _raise                                                                                     │
│                                                                                                  │
│    794 │   │   ex.__cause__ = cause                                                              │
│    795 │   else:                                                                                 │
│    796 │   │   ex.__cause__ = None                                                               │
│ ❱  797 │   raise ex.with_traceback(sys.exc_info()[2])  # set env var OC_CAUSE=1 for full trace   │
│    798                                                                                           │
│    799                                                                                           │
│    800 def format_and_raise(                                                                     │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/omegaconf/listconfig. │
│ py:80 in __init__                                                                                │
│                                                                                                  │
│    77 │   │   │   │   metadata.optional = is_optional                                            │
│    78 │   │   │   │   metadata.element_type = element_type                                       │
│    79 │   │   │   │   self.__dict__["_metadata"] = metadata                                      │
│ ❱  80 │   │   │   self._set_value(value=content, flags=flags)                                    │
│    81 │   │   except Exception as ex:                                                            │
│    82 │   │   │   format_and_raise(node=None, key=key, value=None, cause=ex, msg=str(ex))        │
│    83                                                                                            │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/omegaconf/listconfig. │
│ py:618 in _set_value                                                                             │
│                                                                                                  │
│   615 │   │   except Exception as e:                                                             │
│   616 │   │   │   self.__dict__["_content"] = previous_content                                   │
│   617 │   │   │   self.__dict__["_metadata"] = previous_metadata                                 │
│ ❱ 618 │   │   │   raise e                                                                        │
│   619 │                                                                                          │
│   620 │   def _set_value_impl(                                                                   │
│   621 │   │   self, value: Any, flags: Optional[Dict[str, bool]] = None                          │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/omegaconf/listconfig. │
│ py:614 in _set_value                                                                             │
│                                                                                                  │
│   611 │   │   try:                                                                               │
│   612 │   │   │   previous_content = self.__dict__["_content"]                                   │
│   613 │   │   │   previous_metadata = self.__dict__["_metadata"]                                 │
│ ❱ 614 │   │   │   self._set_value_impl(value, flags)                                             │
│   615 │   │   except Exception as e:                                                             │
│   616 │   │   │   self.__dict__["_content"] = previous_content                                   │
│   617 │   │   │   self.__dict__["_metadata"] = previous_metadata                                 │
│                                                                                                  │
│ /Users/jeremy.dohmann/miniconda3/envs/foundry/lib/python3.10/site-packages/omegaconf/listconfig. │
│ py:646 in _set_value_impl                                                                        │
│                                                                                                  │
│   643 │   │   │   if not (is_primitive_list(value) or isinstance(value, ListConfig)):            │
│   644 │   │   │   │   type_ = type(value)                                                        │
│   645 │   │   │   │   msg = f"Invalid value assigned: {type_.__name__} is not a ListConfig, li   │
│ ❱ 646 │   │   │   │   raise ValidationError(msg)                                                 │
│   647 │   │   │                                                                                  │
│   648 │   │   │   self.__dict__["_content"] = []                                                 │
│   649 │   │   │   if isinstance(value, ListConfig):                                              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ValidationError: Invalid value assigned: generator is not a ListConfig, list or tuple.
    full_key: 
    object_type=None
    ```